File: frontend/src/inference/crisp.ts
// Motor crisp — reglas clásicas (AND de condiciones exactas)
// by Jose Rodas

const OPS: Record<string, (a: any, b: any) => boolean> = {
  '>':  (a, b) => parseFloat(a) > parseFloat(b),
  '<':  (a, b) => parseFloat(a) < parseFloat(b),
  '>=': (a, b) => parseFloat(a) >= parseFloat(b),
  '<=': (a, b) => parseFloat(a) <= parseFloat(b),
  '==': (a, b) => String(a).toLowerCase() === String(b).toLowerCase(),
  '!=': (a, b) => String(a).toLowerCase() !== String(b).toLowerCase(),
};

export function inferirCrisp(
  hechos: Record<string, any>,
  reglas: any[],
): { conclusion: string; certeza: number; explicacion: string }[] {
  const resultados: { conclusion: string; certeza: number; explicacion: string }[] = [];

  for (const r of reglas) {
    const activada = (r.condiciones || []).every(
      (c: any) => OPS[c.op]?.(hechos[c.var] ?? '', c.val) ?? false,
    );
    if (activada) {
      resultados.push({
        conclusion:  r.conclusion,
        certeza:     parseFloat(r.peso ?? 1.0),
        explicacion: r.explicacion ?? '',
      });
    }
  }

  return resultados.sort((a, b) => b.certeza - a.certeza);
}

File: backend/engines/crisp.py
from typing import Any

OPS = {
    ">":  lambda a, b: float(a) > float(b),
    "<":  lambda a, b: float(a) < float(b),
    ">=": lambda a, b: float(a) >= float(b),
    "<=": lambda a, b: float(a) <= float(b),
    "==": lambda a, b: str(a).lower() == str(b).lower(),
    "!=": lambda a, b: str(a).lower() != str(b).lower(),
}

def inferir_crisp(hechos: dict, reglas: list[dict]) -> list[dict]:
    resultados = []
    for r in reglas:
        activada = all(
            OPS.get(c["op"], lambda a, b: False)(hechos.get(c["var"], ""), c["val"])
            for c in r.get("condiciones", [])
        )
        if activada:
            resultados.append({
                "conclusion":  r["conclusion"],
                "certeza":     float(r.get("peso", 1.0)),
                "explicacion": r.get("explicacion", ""),
            })
    return sorted(resultados, key=lambda x: -x["certeza"])
